### PR TITLE
std::cout << "hello OneAPI" << std::endl;

### PR DESCRIPTION
**Pull request by syurkevi**
Adds a OneAPI backend to ArrayFire. This PR is a starting point for further development and collaboration.

Description
-----------

* Is this a new feature or a bug fix?
New feature. New backend!
* More detail if necessary to describe all commits in pull request.
Lots of commits. Commits should probably be squashed. Enough scaffolding is in place to start porting kernels.
* Why these changes are necessary.
This should allow performant support for upcoming Intel hardware.
* Potential impact on specific hardware, software or backends.
This new backend may eventually need to utilize existing JIT implementations. Not sure if changes will be necessary if so.
* Future changes not implemented in this PR.
Sooooo many kernels need to be ported.

Changes to Users
----------------
Current source should remain unaffected when building. If new backend is targeted the icx compiler will be necessary with a command similar to:
```
cmake -DCMAKE_C_COMPILER=icx -DCMAKE_CXX_COMPILER=icx  ..
```

Checklist
---------

- [x] Rebased on latest master
- [x] Code compiles
- [ ] Tests pass
- [ ] Functions added to unified API
- [ ] Functions documented

### CODE REVIEW

File: CMakeModules/InternalUtils.cmake [hunk 1]
```diff
@@ -122,7 +122,7 @@ macro(arrayfire_set_cmake_default_variables)
   set(CMAKE_PREFIX_PATH "${ArrayFire_BINARY_DIR};${CMAKE_PREFIX_PATH}")
   set(BUILD_SHARED_LIBS ON)
 
-  set(CMAKE_CXX_STANDARD 14)
+  set(CMAKE_CXX_STANDARD 17)
```
Comment (9prady9) [inline CMakeModules/InternalUtils.cmake:125]: nice

File: include/af/oneapi.h [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+#include <af/array.h>
+#include <af/dim4.hpp>
+#include <af/exception.h>
+#include <af/device.h>
+#include <stdio.h>
+
+namespace afoneapi
+{
+
+#if 0
```
Comment (9prady9) [inline include/af/oneapi.h:165]: Ah, I missed this one in the first look :)

File: src/backend/common/EventBase.hpp [hunk 3]
```diff
@@ -36,7 +36,8 @@ class EventBase {
 
     /// \brief Event destructor. Calls the destroy event call on the native API
     ~EventBase() noexcept {
-        if (e_) NativeEventPolicy::destroyEvent(&e_);
+      //if (e_)
```
Comment (9prady9) [inline src/backend/common/EventBase.hpp:39]: why leave and comment it out ?

File: src/backend/common/Logger.hpp [hunk 4]
```diff
@@ -13,8 +13,40 @@
...
+/* GNU GCC/G++ */
+#pragma GCC diagnostic pop
+#elif defined(_MSC_VER)
+/* Microsoft Visual Studio */
+#pragma warning(pop)
+#else
+/* Other */
+#endif
+
+
```
Comment (9prady9) [inline src/backend/common/Logger.hpp:49]: This entire pragma based suppression for spdlog is better clubbed in separate header which can be resued from multiple locations in the code base which wants to use spdlog.

File: src/backend/common/forge_loader.hpp [hunk 5]
```diff
@@ -10,10 +10,39 @@
...
+#elif defined(__GNUC__) || defined(__GNUG__)
+/* GNU GCC/G++ */
+#pragma GCC diagnostic pop
+#elif defined(_MSC_VER)
+/* Microsoft Visual Studio */
+#pragma warning(pop)
+#else
+/* Other */
+#endif
+
```
Comment (9prady9) [inline src/backend/common/forge_loader.hpp:45]: Perhaps we can do the same to glad header too (like spdlog)

**Review by 9prady9 — COMMENTED** [review]
Interesting, is this a copy of opencl backend with few fixes and commented out kernel launches ?

File: src/backend/common/forge_loader.hpp (hunk 5 above)
Comment (syurkevi) [reply src/backend/common/forge_loader.hpp:45]: There's only two places it happens, perhaps a macro could work 🤔

**Review by umar456 — DISMISSED** [review]

**Review by umar456 — APPROVED** [review]

**Pull request merged**